Show metadata on connector docs

evantahler opened the pull request:
This PR loads up the latest metadata entry for connectors and injects useful info at the top of the page:


This is accomplished by adding a remark which adds some custom logic to the markdown processor.  We download the latest version of the connector registry at build time and inject a new table at the top of the page showing some handy information:

<img width="1576" alt="Screenshot 2023-11-09 at 4 15 39 PM" src="https://github.com/airbytehq/airbyte/assets/303226/909ddeea-bdf7-49b9-8a2d-9dedeae5a04a">

Pros of this approach:
* No modification of the markdown needed
* No new react components to support in-product

Cons of this approach:
* The registry is downloaded at build-time, and therefore can get stale.  We should periodically rebuild the docs site to overcome this.

bnchrch commented on docusaurus/src/remark/docsHeaderDecoration.js:
```diff
@@ -0,0 +1,135 @@
+
+  loading = true;
+  console.log("Fetching connector registry...");
+  const response = await fetch(registry_url);
```
nit: Why not reuse the same method we have here?

bnchrch commented on docusaurus/src/remark/docsHeaderDecoration.js:
```diff
@@ -0,0 +1,135 @@
+This methods makes registry a singleton available to all callers of this plugin.
+Only one download will happen for the first callers and all others will use the cached version.
+*/
+const fetchCatalog = async () => {
```
nit: Would love if this was hoisted to the top so its right next to the global registry variable

timroes replied:
This whole method seems to be not necessary to me? Can't we simply do inside the `plugin`, but outside the `transform` function a single `fetch` call assign that returning to a local variable and use that everywhere?

i.e. along the lines of

```js
const plugin = () => {
  const registryFetch = fetch("...");
  const transformer = async () => {
    const registry = await registryFetch();

  }
  return transformer;
```

I don't see why we'd need any loading state or similar here?

evantahler replied:
The plugin gets called by every page, so I wanted a global way to block downloading the file more than once, and they all get invoked at the same time

timroes replied:
In this case it should be fine pulling it just out to the top level. Modules are singletons by default in node (why the registry variable above works).

bnchrch commented on docusaurus/src/remark/docsHeaderDecoration.js:
```diff
@@ -0,0 +1,135 @@
+
+const registry_url =
+  "https://connectors.airbyte.com/files/generated_reports/connector_registry_report.json";
+let registry = {};
```
alternative an unessesary idea: Instead of this we could use lodash's [memoize](https://lodash.com/docs/4.17.15#memoize) or [throttle](https://lodash.com/docs/4.17.15#throttle)

timroes replied:
@bnchrch I don't think memoizing or throttling is necessary here. It feels that most of the code here anyway could be shortened away, see my comment in https://github.com/airbytehq/airbyte/pull/32334/files#r1389699474 as well.

bnchrch commented on docusaurus/src/remark/docsHeaderDecoration.js:
```diff
@@ -0,0 +1,135 @@
+  loading = false;
+};
+
+const sleep = (ms) => {
```
👍

bnchrch commented on docusaurus/src/remark/docsHeaderDecoration.js:
```diff
@@ -0,0 +1,135 @@
+  return transformer;
+};
+
+const buildConnectorHTMLContent = (
```
Loved the look of this 😍

bnchrch reviewed (approved):
Great addition. A few nits but I wouldnt stand in the way!

<img src="https://media4.giphy.com/media/R6gvnAxj2ISzJdbA63/giphy-downsized-medium.gif"/>

nataliekwong commented on docusaurus/src/remark/docsHeaderDecoration.js:
```diff
@@ -0,0 +1,135 @@
+          <td>${registryEntry.dockerImageTag_oss}</td>
+        </tr>
+        <tr>
+          <td>Availability:</td>
```
nit (non-blocking): could we re-order these from top to bottom (most relevant at the top):
- Availability
- Support Level
- Latest Version
- Definition ID

nataliekwong reviewed (approved):
Amazing :)

timroes commented on docusaurus/src/remark/docsHeaderDecoration.js:
```diff
@@ -0,0 +1,135 @@
+      <tbody>
+        <tr>
+          <td rowSpan="4">
+            <img style="max-height: 75px" src="${registryEntry.iconUrl_oss}" />
```
Design suggestion: maybe it would look nicer if we just put the icon in front of the title and just have this table contain the facts. It would also render a bit better on mobile, since we only have one column.

timroes commented on docusaurus/src/remark/docsHeaderDecoration.js:
```diff
@@ -0,0 +1,135 @@
+          <td rowSpan="4">
+            <img style="max-height: 75px" src="${registryEntry.iconUrl_oss}" />
+          </td>
+          <td>Support Level: </td>
```
Could we maybe make this (or an additional help icon) a link to the page explaining what those support levels actually mean.

timroes commented on docusaurus/src/remark/docsHeaderDecoration.js:
```diff
@@ -0,0 +1,135 @@
+    if (!registryEntry) return;
+
+    visit(ast, "heading", (node) => {
+      let headerFound = false;
```
This does not work the way you expect. Since this is a block level variable this will always be false in the `if` below. You likely just want to pull the type declaration outside of the visit function. Should likely be easy to test by just adding another H1 in a doc.

timroes reviewed (commented):
Want to test something still later. Please hold back merging yet (in case you wake up before I'm getting to test this).

timroes commented on docusaurus/src/remark/docsHeaderDecoration.js:
```diff
@@ -0,0 +1,135 @@
+*/
+const fetchCatalog = async () => {
+  if (loading) {
+    await sleep(500);
```
What's the purpose for waiting here?

timroes commented on docusaurus/src/remark/docsHeaderDecoration.js:
```diff
@@ -0,0 +1,135 @@
+      </tbody>
+    </table>
+  </small>
+  <br />
```
Manual line breaks are not good HTML practice. The `small` or `table` should just have a classname with a margin-bottom if that's what we're trying to achieve here.

evantahler commented:
<img width="1473" alt="Screenshot 2023-11-10 at 10 18 02 AM" src="https://github.com/airbytehq/airbyte/assets/303226/94c31569-3309-49de-a604-fb445532b459">
I've removed the table and replaced it with a div - I just couldn't get remark to render it without literal whitespaces which the browser complained about

evantahler commented:
Ok, I think this is ready now.  The comments have been addressed.  I've replaced the timer and per-page invocation of `fetchCatalog` with one invocation at the start of the pligin then every transformer just waits until it is done `await Promise.resolve(registryFetcher);`.

Thanks for reminding me how real programming language work, @timroes!

timroes commented on docusaurus/src/remark/docsHeaderDecoration.js:
```diff
@@ -0,0 +1,111 @@
+    const connectorName = pathParts.pop().split(".")[0];
+    const connectorType = pathParts.pop();
+    const dockerRepository = `airbyte/${connectorType.replace(
+      /s$/,
+      ""
```
@evantahler Why are we generically removing any `s` at the end of a connector type?

timroes replied:
Ah I think I get it, this is only here to make `destinations` -> `destination`, `sources` -> `source`.
